### PR TITLE
Don't early terminate on null for 64bit NR HashCode

### PR DESCRIPTION
**Pull request by benaadams**
As used by `NonRandomizedStringEqualityComparer` in `Dictionary<string, TValue>` before it moves to randomized hashing when collisions become high.

Resolves https://github.com/dotnet/coreclr/issues/4681

Merges 32 and 64 bit paths and improves performance ~x2 for longer strings https://github.com/dotnet/coreclr/pull/19331#issuecomment-412242360

Using the 466,544 word from words.txt from https://github.com/dwyl/english-words; and discounting high bit from hashcode (as per Dictionary key)

**Current HashCode**: 90 single collisions, 466364 no collisions
**Improved HashCode**: 119 single collision, 466306 no collisions

```
               Method |                 Input |       Mean | Scaled |
--------------------- |---------------------- |-----------:|-------:|
 Current64BitHashCode |                     1 |   3.037 ns |   1.00 |
     ImprovedHashCode |                     1 |   2.660 ns |   0.88 |
                      |                       |            |        |
 Current64BitHashCode |                    12 |   3.342 ns |   1.00 |
     ImprovedHashCode |                    12 |   2.663 ns |   0.80 |
                      |                       |            |        |
 Current64BitHashCode |                   123 |   3.909 ns |   1.00 |
     ImprovedHashCode |                   123 |   3.189 ns |   0.82 |
                      |                       |            |        |
 Current64BitHashCode |                  1234 |   4.359 ns |   1.00 |
     ImprovedHashCode |                  1234 |   3.183 ns |   0.73 |
                      |                       |            |        |
 Current64BitHashCode |                 12345 |   5.053 ns |   1.00 |
     ImprovedHashCode |                 12345 |   3.489 ns |   0.69 |
                      |                       |            |        |
 Current64BitHashCode |                123456 |   5.453 ns |   1.00 |
     ImprovedHashCode |                123456 |   3.514 ns |   0.64 |
                      |                       |            |        |
 Current64BitHashCode |               1234567 |   6.058 ns |   1.00 |
     ImprovedHashCode |               1234567 |   4.299 ns |   0.71 |
                      |                       |            |        |
 Current64BitHashCode |              12345678 |   6.581 ns |   1.00 |
     ImprovedHashCode |              12345678 |   4.294 ns |   0.65 |
                      |                       |            |        |
 Current64BitHashCode |             123456789 |   7.148 ns |   1.00 |
     ImprovedHashCode |             123456789 |   4.617 ns |   0.65 |
                      |                       |            |        |
 Current64BitHashCode |            1234567890 |   7.693 ns |   1.00 |
     ImprovedHashCode |            1234567890 |   4.619 ns |   0.60 |
                      |                       |            |        |
 Current64BitHashCode |  12345678901234567890 |  13.321 ns |   1.00 |
     ImprovedHashCode |  12345678901234567890 |   7.900 ns |   0.59 |
                      |                       |            |        |
 Current64BitHashCode |  12345(...)67890 [40] |  25.748 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [40] |  13.409 ns |   0.52 |
                      |                       |            |        |
 Current64BitHashCode |  12345(...)67890 [80] |  54.932 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [80] |  25.713 ns |   0.47 |
                      |                       |            |        |
 Current64BitHashCode | 12345(...)67890 [160] | 100.144 ns |   1.00 |
     ImprovedHashCode | 12345(...)67890 [160] |  50.712 ns |   0.51 |
```

@jkotas @GrabYourPitchforks its also used by `RtType.GetHashCodeHelper` https://github.com/dotnet/coreclr/blob/ab9b4511180d1dfde09d1480c29a7bbacf3587dd/src/System.Private.CoreLib/src/System/RtType.cs#L5096-L5104 would this be a concern (e.g changed behaviour for `abcd\0efgh`)

/cc @IDisposable @jamesqo

### CODE REVIEW

**Comment by benaadams** [comment]
Is a question as to why 32 bit operates on `int`s and 64 bit operates on half that `char`s rather than double `long`

**Comment by jkotas** [comment]
The use of RtType.cs is not a concern wrt this change.

The change looks fine to me as is.

I do not think that the exact algorithm matters here. It just needs to good enough and faster than the randomized Marvin. It should be ok to tweak this even further (e.g. vectorize it) if you would like. I would rename the method to GetNonRandomizedHashCode if it stops using the legacy algorithm.

**Comment by benaadams** [comment]
> I do not think that the exact algorithm matters here. It just needs to good enough and faster than the randomized Marvin. It should be ok to tweak this even further (e.g. vectorize it) if you would like.

Ooooh... :)

**Comment by benaadams** [comment]
Definite scope for improvement; though will take some analysis to make sure its not a terrible change (main speed ups are from using longer data type and rotate rather than shift, add, shift)
```
               Method |                 Input |       Mean | Scaled |
--------------------- |---------------------- |-----------:|-------:|
 Current64BitHashCode |                 12345 |   4.723 ns |   1.00 |
 Current32BitHashCode |                 12345 |   3.467 ns |   0.73 |
          NewHashCode |                 12345 |   3.811 ns |   0.81 |
             UintHash |                 12345 |   3.217 ns |   0.68 |
            UlongHash |                 12345 |   3.195 ns |   0.68 |
                      |                       |            |        |
 Current64BitHashCode |              12345678 |   6.279 ns |   1.00 |
 Current32BitHashCode |              12345678 |   4.601 ns |   0.73 |
          NewHashCode |              12345678 |   5.289 ns |   0.84 |
             UintHash |              12345678 |   3.847 ns |   0.61 |
            UlongHash |              12345678 |   4.787 ns |   0.76 |
                      |                       |            |        |
 Current64BitHashCode |            1234567890 |   7.519 ns |   1.00 |
 Current32BitHashCode |            1234567890 |   4.972 ns |   0.66 |
          NewHashCode |            1234567890 |   6.236 ns |   0.83 |
             UintHash |            1234567890 |   4.142 ns |   0.55 |
            UlongHash |            1234567890 |   3.549 ns |   0.47 |
                      |                       |            |        |
 Current64BitHashCode |  12345(...)67890 [40] |  25.484 ns |   1.00 |
 Current32BitHashCode |  12345(...)67890 [40] |  16.555 ns |   0.65 |
          NewHashCode |  12345(...)67890 [40] |  19.649 ns |   0.77 |
             UintHash |  12345(...)67890 [40] |  11.172 ns |   0.44 |
            UlongHash |  12345(...)67890 [40] |   8.466 ns |   0.33 |
                      |                       |            |        |
 Current64BitHashCode | 12345(...)67890 [160] | 104.785 ns |   1.00 |
 Current32BitHashCode | 12345(...)67890 [160] |  64.493 ns |   0.62 |
          NewHashCode | 12345(...)67890 [160] |  83.325 ns |   0.80 |
             UintHash | 12345(...)67890 [160] |  42.317 ns |   0.40 |
            UlongHash | 12345(...)67890 [160] |  26.432 ns |   0.25 |
```

**Comment by benaadams** [comment]
I'll leave it as now and revisit at a later date; the current x64 version is quite good for minimal collisions on a dictionary of 466,544 words; so I'll just add the fix for null values

@dotnet-bot test OSX10.12 x64 Checked CoreFX Tests

**Comment by benaadams** [comment]
Connection refused error on the failed OSX10.12 x64 Checked CoreFX Tests 
```
06:20:21 
   System.Net.Tests.HttpListenerResponseCookiesTests.SetCookie_ValidCookie_ClonesCookie [FAIL]
06:20:21       System.Net.Internals.SocketExceptionFactory+ExtendedSocketException : Connection refused 127.0.0.1:1193
06:20:22       Stack Trace:
06:20:23          /Users/buildagent/agent/_work/350/s/corefx/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(4421,0): at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
06:20:23          /Users/buildagent/agent/_work/350/s/corefx/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(833,0): at System.Net.Sockets.Socket.Connect(EndPoint remoteEP)
06:20:23          /Users/buildagent/agent/_work/350/s/corefx/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(948,0): at System.Net.Sockets.Socket.Connect(IPAddress[] addresses, Int32 port)
06:20:23          --- End of stack trace from previous location where exception was thrown ---
06:20:23          /Users/buildagent/agent/_work/350/s/corefx/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(958,0): at System.Net.Sockets.Socket.Connect(IPAddress[] addresses, Int32 port)
06:20:23          /Users/buildagent/agent/_work/350/s/corefx/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs(902,0): at System.Net.Sockets.Socket.Connect(String host, Int32 port)
06:20:23          /Users/anandono/Source/repos/corefx/src/System.Net.HttpListener/tests/HttpListenerFactory.cs(156,0): at System.Net.Tests.HttpListenerFactory.GetConnectedSocket()
06:20:23          /Users/anandono/Source/repos/corefx/src/System.Net.HttpListener/tests/HttpListenerResponseTests.cs(21,0): at System.Net.Tests.HttpListenerResponseTestBase..ctor(
```

Accidentally started a  bunch of other ones, but didn't restart that one 😢

**Comment by benaadams** [comment]
I tried changing it to an `pointer[index]` and length check rather than increment the pointer, but the more complex address lookup slowed it down :-/

Seems a bit senstive

**Comment by benaadams** [comment]
Some kind of build issue?
```
20:36:59 WARNING: System.IO.Tests.FileStream_SafeFileHandle.ThrowWhenHandlePositionIsChanged_sync is running for 4339s.
20:42:20 FATAL: command execution failed
20:42:20 java.nio.channels.ClosedChannelException
20:42:20 	at org.jenkinsci.remoting.protocol.impl.ChannelApplicationLayer.onReadClosed(ChannelApplicationLayer.java:208)
20:42:20 	at org.jenkinsci.remoting.protocol.ApplicationLayer.onRecvClosed(ApplicationLayer.java:222)
20:42:20 	at org.jenkinsci.remoting.protocol.ProtocolStack$Ptr.onRecvClosed(ProtocolStack.java:832)
20:42:20 	at org.jenkinsci.remoting.protocol.FilterLayer.onRecvClosed(FilterLayer.java:287)
20:42:20 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.onRecvClosed(SSLEngineFilterLayer.java:181)
20:42:20 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.switchToNoSecure(SSLEngineFilterLayer.java:283)
20:42:20 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.processWrite(SSLEngineFilterLayer.java:503)
20:42:20 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.processQueuedWrites(SSLEngineFilterLayer.java:248)
20:42:20 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.doSend(SSLEngineFilterLayer.java:200)
20:42:20 	at org.jenkinsci.remoting.protocol.impl.SSLEngineFilterLayer.doCloseSend(SSLEngineFilterLayer.java:213)
20:42:20 	at org.jenkinsci.remoting.protocol.ProtocolStack$Ptr.doCloseSend(ProtocolStack.java:800)
20:42:20 	at org.jenkinsci.remoting.protocol.ApplicationLayer.doCloseWrite(ApplicationLayer.java:173)
20:42:20 	at org.jenkinsci.remoting.protocol.impl.ChannelApplicationLayer$ByteBufferCommandTransport.closeWrite(ChannelApplicationLayer.java:311)
20:42:20 	at hudson.remoting.Channel.close(Channel.java:1405)
20:42:20 	at hudson.remoting.Channel.close(Channel.java:1358)
20:42:20 	at hudson.slaves.SlaveComputer.closeChannel(SlaveComputer.java:737)
20:42:20 	at hudson.slaves.SlaveComputer.access$800(SlaveComputer.java:96)
20:42:20 	at hudson.slaves.SlaveComputer$3.run(SlaveComputer.java:655)
20:42:20 	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
20:42:20 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
20:42:20 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
20:42:20 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
20:42:20 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
20:42:20 	at java.lang.Thread.run(Thread.java:748)
20:42:20 Caused: hudson.remoting.ChannelClosedException: Remote call on JNLP4-connect connection from 40.112.192.84/40.112.192.84:49176 failed. The channel is closing down or has closed down
20:42:20 	at hudson.remoting.Channel.call(Channel.java:901)
20:42:20 	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:281)
```

**Comment by benaadams** [comment]
Using the 466,544 word from words.txt from https://github.com/dwyl/english-words; and discounting high bit from hashcode (as per Dictionary key)

**Current HashCode**: 90 single collisions, 466364 no collisions
**Improved HashCode**: 119 single collision, 466306 no collisions

```
               Method |                 Input |       Mean | Scaled |          Op/s |
--------------------- |---------------------- |-----------:|-------:|--------------:|
 Current64BitHashCode |                     1 |   3.770 ns |   1.00 | 265,251,989.4 |
     ImprovedHashCode |                     1 |   3.123 ns |   0.83 | 320,204,931.6 |
                      |                       |            |        |               |
 Current64BitHashCode |                    12 |   3.159 ns |   1.00 | 316,555,872.1 |
     ImprovedHashCode |                    12 |   3.677 ns |   1.16 | 271,960,837.6 |
                      |                       |            |        |               |
 Current64BitHashCode |                   123 |   5.582 ns |   1.00 | 179,147,259.0 |
     ImprovedHashCode |                   123 |   4.237 ns |   0.76 | 236,016,049.1 |
                      |                       |            |        |               |
 Current64BitHashCode |                  1234 |   4.299 ns |   1.00 | 232,612,235.4 |
     ImprovedHashCode |                  1234 |   3.955 ns |   0.92 | 252,844,500.6 |
                      |                       |            |        |               |
 Current64BitHashCode |                 12345 |   6.439 ns |   1.00 | 155,303,618.6 |
     ImprovedHashCode |                 12345 |   4.239 ns |   0.66 | 235,904,694.5 |
                      |                       |            |        |               |
 Current64BitHashCode |                123456 |   5.415 ns |   1.00 | 184,672,206.8 |
     ImprovedHashCode |                123456 |   4.233 ns |   0.78 | 236,239,073.9 |
                      |                       |            |        |               |
 Current64BitHashCode |               1234567 |   7.276 ns |   1.00 | 137,438,152.8 |
     ImprovedHashCode |               1234567 |   4.521 ns |   0.62 | 221,190,002.2 |
                      |                       |            |        |               |
 Current64BitHashCode |              12345678 |   6.532 ns |   1.00 | 153,092,467.9 |
     ImprovedHashCode |              12345678 |   5.083 ns |   0.78 | 196,734,212.1 |
                      |                       |            |        |               |
 Current64BitHashCode |             123456789 |   8.432 ns |   1.00 | 118,595,825.4 |
     ImprovedHashCode |             123456789 |   5.387 ns |   0.64 | 185,632,077.2 |
                      |                       |            |        |               |
 Current64BitHashCode |            1234567890 |   7.674 ns |   1.00 | 130,310,138.1 |
     ImprovedHashCode |            1234567890 |   5.355 ns |   0.70 | 186,741,363.2 |
                      |                       |            |        |               |
 Current64BitHashCode |  12345678901234567890 |  13.326 ns |   1.00 |  75,041,272.7 |
     ImprovedHashCode |  12345678901234567890 |   8.448 ns |   0.63 | 118,371,212.1 |
                      |                       |            |        |               |
 Current64BitHashCode |  12345(...)67890 [40] |  25.967 ns |   1.00 |  38,510,417.1 |
     ImprovedHashCode |  12345(...)67890 [40] |  14.167 ns |   0.55 |  70,586,574.4 |
                      |                       |            |        |               |
 Current64BitHashCode |  12345(...)67890 [80] |  54.985 ns |   1.00 |  18,186,778.2 |
     ImprovedHashCode |  12345(...)67890 [80] |  25.409 ns |   0.46 |  39,356,133.7 |
                      |                       |            |        |               |
 Current64BitHashCode | 12345(...)67890 [160] | 100.555 ns |   1.00 |   9,944,806.3 |
     ImprovedHashCode | 12345(...)67890 [160] |  51.937 ns |   0.52 |  19,254,096.3 |
```

**Comment by benaadams** [comment]
Works better than System.HashCode for the reduced input domain of strings
```
              Method | No Collisions |  1 Coll |  2 Coll |  3 Coll |  4 Coll |  5 Coll |
-------------------- |-------------- |-------- |-------- |-------- |-------- |-------- |
Current64BitHashCode |        466364 |      90 |       - |       - |       - |       - |
    ImprovedHashCode |        466306 |     119 |       - |       - |       - |       - |
      SystemHashCode |        453811 |    5575 |     431 |      56 |      12 |       1 |
```
```
               Method |                 Input |       Mean | Scaled |
--------------------- |---------------------- |-----------:|-------:|
 Current64BitHashCode |                     1 |   3.056 ns |   1.00 |
     ImprovedHashCode |                     1 |   3.047 ns |   1.00 |
       SystemHashCode |                     1 |  18.065 ns |   5.91 |
                      |                       |            |        |
 Current64BitHashCode |                    12 |   3.411 ns |   1.00 |
     ImprovedHashCode |                    12 |   3.413 ns |   1.00 |
       SystemHashCode |                    12 |  17.987 ns |   5.27 |
                      |                       |            |        |
 Current64BitHashCode |                   123 |   4.014 ns |   1.00 |
     ImprovedHashCode |                   123 |   3.945 ns |   0.98 |
       SystemHashCode |                   123 |  22.657 ns |   5.64 |
                      |                       |            |        |
 Current64BitHashCode |                  1234 |   4.552 ns |   1.00 |
     ImprovedHashCode |                  1234 |   3.656 ns |   0.80 |
       SystemHashCode |                  1234 |  23.885 ns |   5.25 |
                      |                       |            |        |
 Current64BitHashCode |                 12345 |   5.117 ns |   1.00 |
     ImprovedHashCode |                 12345 |   4.168 ns |   0.81 |
       SystemHashCode |                 12345 |  28.089 ns |   5.49 |
                      |                       |            |        |
 Current64BitHashCode |                123456 |   5.659 ns |   1.00 |
     ImprovedHashCode |                123456 |   3.937 ns |   0.70 |
       SystemHashCode |                123456 |  24.214 ns |   4.28 |
                      |                       |            |        |
 Current64BitHashCode |               1234567 |   6.269 ns |   1.00 |
     ImprovedHashCode |               1234567 |   4.260 ns |   0.68 |
       SystemHashCode |               1234567 |  28.053 ns |   4.47 |
                      |                       |            |        |
 Current64BitHashCode |              12345678 |   6.788 ns |   1.00 |
     ImprovedHashCode |              12345678 |   4.839 ns |   0.71 |
       SystemHashCode |              12345678 |  27.865 ns |   4.10 |
                      |                       |            |        |
 Current64BitHashCode |             123456789 |   7.377 ns |   1.00 |
     ImprovedHashCode |             123456789 |   5.340 ns |   0.72 |
       SystemHashCode |             123456789 |  34.938 ns |   4.74 |
                      |                       |            |        |
 Current64BitHashCode |            1234567890 |   7.982 ns |   1.00 |
     ImprovedHashCode |            1234567890 |   5.569 ns |   0.70 |
       SystemHashCode |            1234567890 |  28.755 ns |   3.60 |
                      |                       |            |        |
 Current64BitHashCode |  12345678901234567890 |  13.531 ns |   1.00 |
     ImprovedHashCode |  12345678901234567890 |   8.225 ns |   0.61 |
       SystemHashCode |  12345678901234567890 |  42.936 ns |   3.17 |
                      |                       |            |        |
 Current64BitHashCode |  12345(...)67890 [40] |  25.915 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [40] |  13.884 ns |   0.54 |
       SystemHashCode |  12345(...)67890 [40] |  63.722 ns |   2.46 |
                      |                       |            |        |
 Current64BitHashCode |  12345(...)67890 [80] |  51.144 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [80] |  25.459 ns |   0.50 |
       SystemHashCode |  12345(...)67890 [80] | 120.161 ns |   2.35 |
                      |                       |            |        |
 Current64BitHashCode | 12345(...)67890 [160] |  99.804 ns |   1.00 |
     ImprovedHashCode | 12345(...)67890 [160] |  49.764 ns |   0.50 |
       SystemHashCode | 12345(...)67890 [160] | 215.416 ns |   2.16 |
```

File: src/System.Private.CoreLib/shared/System/String.Comparison.cs [hunk 1]
```diff
@@ -755,52 +755,62 @@ public override int GetHashCode()
 
         // Use this if and only if you need the hashcode to not change across app domains (e.g. you have an app domain agile
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/shared/System/String.Comparison.cs]: I guess this comment should be replaced with a note to only use it if DOS is not a concern or is otherwise mitigated?

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/String.Comparison.cs]: Updated to:
```
// Use this if and only if 'Denial of Service' attacks are not a concern (i.e. never used for free-form user input),
// or are otherwise mitigated
```

**Comment by danmosemsft** [comment]
Curious how `string.GetHashCode()` (which is Marvin) compares to the System.HashCode you measured above -- although neither are material to this change of course.

**Comment by danmosemsft** [comment]
Your measurements above are all for 64 bit, but the 32 bit path is changed also, right? Should we check that isn't slower?

**Comment by danmosemsft** [comment]
I would like to see 32 bit perf numbers to check if possible. (Assuming I read correctly that these are all 64 bit perf numbers)

> It might be worth investigating processing doing more in the long space on x64 bit; as this is moving to int from `char and may pay back on shorter (and more common) strings.

Can you clarify - I see you are already reading a ulong at a time on 64 bit?

@jkotas, you reviewed already, any remaining concerns?

**Comment by benaadams** [comment]
32 bit versions; should I ifdef back to what it was previously on 32bit? (maybe 32bit Jit doesn't recognise rotate)
```
               Method |                 Input |      Mean | Scaled |
--------------------- |---------------------- |----------:|-------:|
 Current32BitHashCode |                     1 |  3.091 ns |   1.00 |
     ImprovedHashCode |                     1 |  3.769 ns |   1.22 |
                      |                       |           |        |
 Current32BitHashCode |                    12 |  3.065 ns |   1.00 |
     ImprovedHashCode |                    12 |  4.526 ns |   1.48 |
                      |                       |           |        |
 Current32BitHashCode |                   123 |  3.949 ns |   1.00 |
     ImprovedHashCode |                   123 |  4.822 ns |   1.22 |
                      |                       |           |        |
 Current32BitHashCode |                  1234 |  3.952 ns |   1.00 |
     ImprovedHashCode |                  1234 |  4.833 ns |   1.22 |
                      |                       |           |        |
 Current32BitHashCode |                 12345 |  4.221 ns |   1.00 |
     ImprovedHashCode |                 12345 |  5.106 ns |   1.21 |
                      |                       |           |        |
 Current32BitHashCode |                123456 |  4.267 ns |   1.00 |
     ImprovedHashCode |                123456 |  5.141 ns |   1.21 |
                      |                       |           |        |
 Current32BitHashCode |               1234567 |  5.475 ns |   1.00 |
     ImprovedHashCode |               1234567 |  5.336 ns |   0.97 |
                      |                       |           |        |
 Current32BitHashCode |              12345678 |  5.460 ns |   1.00 |
     ImprovedHashCode |              12345678 |  5.916 ns |   1.08 |
                      |                       |           |        |
 Current32BitHashCode |             123456789 |  5.621 ns |   1.00 |
     ImprovedHashCode |             123456789 |  6.641 ns |   1.18 |
                      |                       |           |        |
 Current32BitHashCode |            1234567890 |  5.614 ns |   1.00 |
     ImprovedHashCode |            1234567890 |  6.502 ns |   1.16 |
                      |                       |           |        |
 Current32BitHashCode |  12345678901234567890 |  9.788 ns |   1.00 |
     ImprovedHashCode |  12345678901234567890 | 10.554 ns |   1.08 |
                      |                       |           |        |
 Current32BitHashCode |  12345(...)67890 [40] | 16.944 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [40] | 17.873 ns |   1.05 |
                      |                       |           |        |
 Current32BitHashCode |  12345(...)67890 [80] | 31.119 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [80] | 33.047 ns |   1.06 |
                      |                       |           |        |
 Current32BitHashCode | 12345(...)67890 [160] | 67.350 ns |   1.00 |
     ImprovedHashCode | 12345(...)67890 [160] | 66.761 ns |   0.99 |
```

> Can you clarify - I see you are already reading a ulong at a time on 64 bit?

It splits the long and processes it as 2 `int`s for `hash1` and `hash2`. It could in theory process it as longs instead; however the maths would change; and need more validation to ensure it was still good; I went the easier route 😄.

File: src/System.Private.CoreLib/shared/System/String.Comparison.cs [hunk 2]
```diff
@@ -753,54 +753,64 @@ public override int GetHashCode()
...
+
+                int length = this.Length;
+                uint* current = (uint*)src;
+
+                char* end = (src + length);
 #if BIT64
-                    int hash1 = 5381;
-#else // !BIT64 (32)
-                    int hash1 = (5381<<16) + 5381;
+                if (length >= 4 && (((int)current) & 7) != 0)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/String.Comparison.cs]: `(((int)current) & 7) != 0` is always going to be true given the current String layout on 64-bit.

**Comment by jkotas** [comment]
LGTM modulo the nits

**Comment by benaadams** [comment]
> 32 bit versions; should I ifdef back to what it was previously on 32bit? (maybe 32bit Jit doesn't recognise rotate)

Ahh... I didn't test with the 64bit stuff commented out...

File: src/System.Private.CoreLib/shared/System/String.Comparison.cs (hunk 2 above)
Comment (danmosemsft) [reply src/System.Private.CoreLib/shared/System/String.Comparison.cs]: Maybe should be a Debug.Assert instead?

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/String.Comparison.cs]: I'm not sure that's true? It seems interned strings sizes 0 - 3 go down the other path?

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/String.Comparison.cs]: duh; but I test that the length is `>= 4` first...

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/String.Comparison.cs]: oh, that's probably why its going down the other path; ok, bit embarrassed...

**Comment by benaadams** [comment]
32bit previously skips the extra test by consuming the null terminator; but also ensures its there in the Debug test (e.g. odd lengths consume the null; even lengths don't) so I'm loosing some speed from that.

Will also consume the null of odd lengths.

**Comment by benaadams** [comment]
Now its a bit simpler and uses the 32bit variant for both; but uses the rotate pattern and is a bit more streamlined.

32 bit results
```
BenchmarkDotNet=v0.11.0, OS=Windows 10.0.17733
Intel Core i7-4720HQ CPU 2.60GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.302
  [Host]     : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 32bit RyuJIT
  32 bit cli : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 32bit RyuJIT

Job=32 bit cli  Jit=RyuJit  Platform=X86
Toolchain=32 bit cli

               Method |                 Input |      Mean | Scaled |
--------------------- |---------------------- |----------:|-------:|
 Current32BitHashCode |                     1 |  2.941 ns |   1.00 |
     ImprovedHashCode |                     1 |  2.909 ns |   0.99 |
                      |                       |           |        |
 Current32BitHashCode |                    12 |  3.026 ns |   1.00 |
     ImprovedHashCode |                    12 |  2.901 ns |   0.96 |
                      |                       |           |        |
 Current32BitHashCode |                   123 |  3.785 ns |   1.00 |
     ImprovedHashCode |                   123 |  3.506 ns |   0.93 |
                      |                       |           |        |
 Current32BitHashCode |                  1234 |  3.783 ns |   1.00 |
     ImprovedHashCode |                  1234 |  3.501 ns |   0.93 |
                      |                       |           |        |
 Current32BitHashCode |                 12345 |  4.080 ns |   1.00 |
     ImprovedHashCode |                 12345 |  3.756 ns |   0.92 |
                      |                       |           |        |
 Current32BitHashCode |                123456 |  4.080 ns |   1.00 |
     ImprovedHashCode |                123456 |  3.758 ns |   0.92 |
                      |                       |           |        |
 Current32BitHashCode |               1234567 |  5.316 ns |   1.00 |
     ImprovedHashCode |               1234567 |  4.354 ns |   0.82 |
                      |                       |           |        |
 Current32BitHashCode |              12345678 |  5.298 ns |   1.00 |
     ImprovedHashCode |              12345678 |  4.347 ns |   0.82 |
                      |                       |           |        |
 Current32BitHashCode |             123456789 |  5.507 ns |   1.00 |
     ImprovedHashCode |             123456789 |  4.632 ns |   0.84 |
                      |                       |           |        |
 Current32BitHashCode |            1234567890 |  5.465 ns |   1.00 |
     ImprovedHashCode |            1234567890 |  4.613 ns |   0.84 |
                      |                       |           |        |
 Current32BitHashCode |  12345678901234567890 |  9.575 ns |   1.00 |
     ImprovedHashCode |  12345678901234567890 |  7.189 ns |   0.75 |
                      |                       |           |        |
 Current32BitHashCode |  12345(...)67890 [40] | 16.573 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [40] | 11.921 ns |   0.72 |
                      |                       |           |        |
 Current32BitHashCode |  12345(...)67890 [80] | 30.751 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [80] | 21.955 ns |   0.71 |
                      |                       |           |        |
 Current32BitHashCode | 12345(...)67890 [160] | 65.396 ns |   1.00 |
     ImprovedHashCode | 12345(...)67890 [160] | 45.782 ns |   0.70 |
```

**Comment by benaadams** [comment]
64 bit results
```
BenchmarkDotNet=v0.11.0, OS=Windows 10.0.17733
Intel Core i7-4720HQ CPU 2.60GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.400-preview-009171
  [Host]     : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 64bit RyuJIT


               Method |                 Input |       Mean | Scaled |
--------------------- |---------------------- |-----------:|-------:|
 Current64BitHashCode |                     1 |   3.037 ns |   1.00 |
     ImprovedHashCode |                     1 |   2.660 ns |   0.88 |
                      |                       |            |        |
 Current64BitHashCode |                    12 |   3.342 ns |   1.00 |
     ImprovedHashCode |                    12 |   2.663 ns |   0.80 |
                      |                       |            |        |
 Current64BitHashCode |                   123 |   3.909 ns |   1.00 |
     ImprovedHashCode |                   123 |   3.189 ns |   0.82 |
                      |                       |            |        |
 Current64BitHashCode |                  1234 |   4.359 ns |   1.00 |
     ImprovedHashCode |                  1234 |   3.183 ns |   0.73 |
                      |                       |            |        |
 Current64BitHashCode |                 12345 |   5.053 ns |   1.00 |
     ImprovedHashCode |                 12345 |   3.489 ns |   0.69 |
                      |                       |            |        |
 Current64BitHashCode |                123456 |   5.453 ns |   1.00 |
     ImprovedHashCode |                123456 |   3.514 ns |   0.64 |
                      |                       |            |        |
 Current64BitHashCode |               1234567 |   6.058 ns |   1.00 |
     ImprovedHashCode |               1234567 |   4.299 ns |   0.71 |
                      |                       |            |        |
 Current64BitHashCode |              12345678 |   6.581 ns |   1.00 |
     ImprovedHashCode |              12345678 |   4.294 ns |   0.65 |
                      |                       |            |        |
 Current64BitHashCode |             123456789 |   7.148 ns |   1.00 |
     ImprovedHashCode |             123456789 |   4.617 ns |   0.65 |
                      |                       |            |        |
 Current64BitHashCode |            1234567890 |   7.693 ns |   1.00 |
     ImprovedHashCode |            1234567890 |   4.619 ns |   0.60 |
                      |                       |            |        |
 Current64BitHashCode |  12345678901234567890 |  13.321 ns |   1.00 |
     ImprovedHashCode |  12345678901234567890 |   7.900 ns |   0.59 |
                      |                       |            |        |
 Current64BitHashCode |  12345(...)67890 [40] |  25.748 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [40] |  13.409 ns |   0.52 |
                      |                       |            |        |
 Current64BitHashCode |  12345(...)67890 [80] |  54.932 ns |   1.00 |
     ImprovedHashCode |  12345(...)67890 [80] |  25.713 ns |   0.47 |
                      |                       |            |        |
 Current64BitHashCode | 12345(...)67890 [160] | 100.144 ns |   1.00 |
     ImprovedHashCode | 12345(...)67890 [160] |  50.712 ns |   0.51 |
```

**Comment by benaadams** [comment]
Collisions check
```
              Method | No Collisions |  1 Coll |  2 Coll |
-------------------- |-------------- |-------- |-------- |
Current64BitHashCode |        466364 |      90 |       - |
Current32BitHashCode |        466388 |      78 |       - |
    ImprovedHashCode |        466339 |     101 |       1 |
```

**Comment by benaadams** [comment]
The change in collisions between current 32bit version and new is previously it used a sign extending left shift + right shift; and now it is unsigned (rotate).

**Comment by danmosemsft** [comment]
Thanks @benaadams, nice improvement!
